src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -416,6 +432,70 @@ private Object readResolve() {
             return getDisplayName();
         }
 
+        /** hash code of list of heads */
+        private transient int lastCheckedHashCode;
+        private transient String lastEnclosingLabel;
+        @Restricted(NoExternalUse.class) // for Jelly
+        public @CheckForNull String getEnclosingLabel() {
```

Comment:
I would synchronize this method or offload it to the async operation (via AJAX). It includes an operation with a 100 ms timeout, hence the current implementation may consume all Jenkins executors if there are many ExecutorStep rendering requests